Require Java 11 or newer (#382)
jenkinsci/jep: jep/0000/README.adoc

    link:https://www.jenkins.io/security/#team[Security team]'s instance
* Maven build flow (
    link:https://github.com/jenkinsci/maven-hpi-plugin[Maven HPI Plugin],
    link:https://github.com/jenkinsci/plugin-pom[Plugin POM],
    link:https://github.com/jenkinsci/pom[Jenkins POM],
    link:https://github.com/jenkinsci/bom[Jenkins plugin bill of materials POM],
    etc.)
* Test tools:
    link:https://github.com/jenkinsci/jenkins-test-harness[Jenkins Test Harness (JTH)],
    link:https://github.com/jenkinsci/acceptance-test-harness[Acceptance Test Harness (ATH)],
    link:https://github.com/jenkinsci/plugin-compat-tester[Plugin Compatibility Tester (PCT)]
* User and Developer Documentation

Out of the scope for this JEP:

* Packaging in subprojects (unless specifically noted): Jenkins Operator, Jenkins Evergreen, Jenkinsfile Runner, etc.
  They will be handled in follow-up JEPs if needed.
* Gradle build flow

=== Jenkins core patches

Work to be considered is defined in link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688].

==== Library updates

* The link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688 epic] will include library updates as needed.

> * Some updates may require downstream plugin updates.

Reviewer: noting @jglick's suggestion in https://github.com/jenkinsci/remoting/pull/507#issuecomment-1030629619

may be a better place for it but just recording the feedback:
```suggestion
* Some updates may require downstream plugin updates.
* JavaWS support to be removed from remoting - link:https://github.com/jenkinsci/remoting/pull/507#issuecomment-1030629619[jenkinsci/remoting#507]
```

==== Core patches

* link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core
* link:https://github.com/jenkinsci/jenkins/pull/6086[PR 6086] - Switch Jenkins acceptance test harness in core to Java 11
* link:https://github.com/jenkinsci/jenkins/pull/6092[PR 6092] - Add administrative monitor to remind of upcoming end of Java 8 support

==== Build flow updates

* Jenkinsfile is updated to stop running tests with Java 8
** It includes Unit tests, JTH and ATH smoke tests
* It is possible to build Jenkins Core with the release profile on Java 8

=== Jenkins Docker packaging

The containers tagged for Java 8, like `latest-jdk8` and `centos7-jdk8` will no longer be updated.
The upgrade guide and the announcement blogpost will note that users must switch to other images.
Labels will not be removed for existing containers, but those labels will not be provided for new builds.

Java 8 images will no longer be provided for the controller containers:

* https://hub.docker.com/r/jenkins/jenkins[Controller]

==== Jenkins Agent Images

> Java 8 images will no longer be provided for the general purpose agent containers:

Reviewer 2: What's going to be the policy for compatibility with agents? Would a controller running on Java 11 work with an agent running on Java 8? IOW once Java 11 APIs can be used in Jenkins core, are they going to be allowed in classes that are known to be serialized to agents?

> = JEP-0000: End Java 8 support in Jenkins

Reviewer 3: I do not think the phrasing in the title is clear enough. It makes it sound like we will close bug reports related to Java 8 as “will not fix” but you could still kind of keep things going if you did not want “support”. That is not what is being proposed here IIUC, or at least not what the key https://github.com/jenkinsci/jenkins/pull/6083 does: it means that subsequent Jenkins releases will not run on Java 8, at all. `/opt/jdk8/bin/java -war jenkins.war` will just print an error about unrecognized bytecode version and immediately exit. So it would be clearer to emphasize that the change is to require Java 11 (or newer I suppose) to run.

(IIRC previous Java upgrades involved some patches to `executable-war` and/or `remoting` to compile the entry point class with an older bytecode version so that it could detect the JRE version and print a nicer error.)

Also BTW the PR title (but not the JEP title) talks about Jenkins _core_, which is accurate only to a certain extent—plugins may continue to “support” Java 8 indefinitely, but only insofar as they avoid updating their core baseline (and perhaps also avoid updating the parent POM, TBD).

Author: I like the insight that the title is not clear enough.  Here are some alternatives for consideration

State the new minimum Java version instead of stating the end of support of Java 8

= JEP-0000: Require Java 11 as minimum Java version

Alternatively, a shorter form of the same idea

= JEP-0000: Require Java 11 or newer

Another alternative, state the technical change

= JEP-0000: Use Java 11 bytecode instead of Java 8 bytecode

I'm open to other alternatives as well.

Reviewer 3: > Use Java 11 bytecode instead of Java 8 bytecode

Misleading, because we are also requiring Java 11 API surface, not merely bytecode.

> Require Java 11 or newer

I like this. Succinct and emphatic.

Author: Fixed title in 2acc2f6f0a7d1c5889a98ce13d272c15495a38e6

Author: Fixed in 718a3ac79b2b2ffc7ad68e59dff5c1f9a7fa09fe

| BDFL-Delegate
| TBD

| Pull requests
a|

- link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core
- link:https://github.com/jenkinsci/jenkins/pull/6086[PR 6086] - Switch Jenkins acceptance test harness in core to Java 11
- link:https://github.com/jenkinsci/jenkins/pull/6092[PR 6092] - Add administrative monitor to remind of upcoming end of Java 8 support
- link:https://github.com/jenkinsci/acceptance-test-harness/pull/726[PR 726] - Drop Java 8 from acceptance test harness

| Discussions-To
| link:https://groups.google.com/g/jenkinsci-dev[Jenkins Developer Mailing List]

//
// Uncomment if this JEP depends on one or more other JEPs.
//| Requires
//| :bulb: JEP-NUMBER, JEP-NUMBER... :bulb:
//
//
// Uncomment and fill if this JEP is rendered obsolete by a later JEP
//| Superseded-By
//| :bulb: JEP-NUMBER :bulb:
//
//
// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
//| Resolution
//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:

|===

== Abstract

Java 11 was released on September 25, 2018.
It is an LTS version with a long support timeline.
Java 8 was released in March 2014.

> The link:https://www.oracle.com/java/technologies/java-se-support-roadmap.html[Oracle Java SE Support Roadmap] states that premier support for Java 8 ends in March 2022.

Reviewer: yeah that's what i'm worried about putting this in a reasoning, it will give people who are against this an argument, well my vendor is supporting it till 2030...

> The link:https://www.oracle.com/java/technologies/java-se-support-roadmap.html[Oracle Java SE Support Roadmap] states that premier support for Java 8 ends in March 2022.

Reviewer 3: If there is a legitimate argument then it should be discussed in a section in **Reasoning**, with an explanation of why we chose this approach nonetheless.

> The link:https://www.oracle.com/java/technologies/java-se-support-roadmap.html[Oracle Java SE Support Roadmap] states that premier support for Java 8 ends in March 2022.

Author: Points well taken.  Removed in 5dec80af82a1e679f587a3139303fcf0598de9d9

Java 8 enhancements have stopped as effort focuses on Java 11, Java 17, and beyond.

This Jenkins Enhancement Proposal describes actions required
to make Jenkins end its support of Java 8.

== Specification

The specification has been created as a result of a link:https://groups.google.com/g/jenkinsci-dev/c/YghQ0YP4m78/m/LO9AFa_GAgAJ[Jenkins developers mailing list discussion].

All stories in this JEP have tickets created in Jira or pull requests in GitHub.
Jira tickets are aggregated in
link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688].

> If the tickets are not linked in the description, they can be found there.

Reviewer 3: What does this sentence mean? Description of what? Where is “there”?

Author: Removed

=== Goals and non-goals

Goals:

* Jenkins core development requires Java 11 or newer

> * Jenkins plugin development requires Java 11 or newer when the plugin updates its minimum Java version to a Jenkins core version that does not support Java 8

Reviewer 3: ```suggestion
* Jenkins plugin development requires Java 11 or newer when the plugin updates its minimum Jenkins core version to one that does not support Java 8
```

Non-goals:

* Building all components with Java 11

> * Jenkins replace older API calls with Java 11 API calls

Reviewer 3: ```suggestion
* Jenkins replacing older API calls with Java 11 API calls
```

(non-parallel grammatical structures in bullet points, just awkward)

** Jakarta EE upgrade is not required as part of this JEP
* All plugins are operational with Java 11

> ** Some Jenkins components may be updated to support features offered in Java 11 and beyond, but there is no plan to update all tools

Reviewer 3: Which “components” and “tools” are being discussed here? Is this deliberately a sub-bullet of the one discussing plugins?

> * Multi-release jar support in development tools

Reviewer 3: ```suggestion
* Multi-release JAR support in development tools
```
Good since IMO we should not use multi-release JARs ever.

* Cleanup of removed or deprecated features

=== Scope of changes

* Jenkins core
* Jenkins remoting

> * Jenkins configuration as code

Reviewer 3: This is just a plugin. Why is it listed specially?

* Docker packaging
* Plugins
* Jenkins website: www.jenkins.io
* Jenkins CI instances:

>     link:https://ci.jenkins.io/[ci.jenkins.io],

Reviewer 3: Already running on 11 AFAIK?

    infra.ci.jenkins.io,
    trusted.ci.jenkins.io,
    link:https://www.jenkins.io/security/#team[Security team]'s instance

> * Maven build flow (
>     link:https://github.com/jenkinsci/maven-hpi-plugin[Maven HPI Plugin],

Reviewer 3: ```suggestion
* Maven build flow (link:https://github.com/jenkinsci/maven-hpi-plugin[Maven HPI Plugin],
```

    link:https://github.com/jenkinsci/plugin-pom[Plugin POM],
    link:https://github.com/jenkinsci/pom[Jenkins POM],
    link:https://github.com/jenkinsci/bom[Jenkins plugin bill of materials POM],
    etc.)
* Test tools:
    link:https://github.com/jenkinsci/jenkins-test-harness[Jenkins Test Harness (JTH)],
    link:https://github.com/jenkinsci/acceptance-test-harness[Acceptance Test Harness (ATH)],
    link:https://github.com/jenkinsci/plugin-compat-tester[Plugin Compatibility Tester (PCT)]
* User and Developer Documentation

Out of the scope for this JEP:

* Packaging in subprojects (unless specifically noted): Jenkins Operator, Jenkins Evergreen, Jenkinsfile Runner, etc.

>   They will be handled in follow-up JEPs if needed.

Reviewer 3: Or just adapted without JEPs, presumably.

* Gradle build flow

=== Jenkins core patches

Work to be considered is defined in link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688].

==== Library updates

* The link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688 epic] will include library updates as needed.
* Some updates may require downstream plugin updates.

==== Core patches

* link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core
* link:https://github.com/jenkinsci/jenkins/pull/6086[PR 6086] - Switch Jenkins acceptance test harness in core to Java 11
* link:https://github.com/jenkinsci/jenkins/pull/6092[PR 6092] - Add administrative monitor to remind of upcoming end of Java 8 support

==== Build flow updates

> * Jenkinsfile is updated to stop running tests with Java 8

Reviewer 3: More significantly, it is built using Java 11 bytecode and potentially using Java 11+ APIs.

** It includes Unit tests, JTH and ATH smoke tests

> * It is possible to build Jenkins Core with the release profile on Java 8

Reviewer 3: Not according to https://github.com/jenkinsci/jenkins/pull/6083.

Java 8 images will no longer be provided for the controller containers:

* https://hub.docker.com/r/jenkins/jenkins[Controller]

==== Jenkins Agent Images

Java 8 images will no longer be provided for the general purpose agent containers:

* https://hub.docker.com/r/jenkins/agent[Agent]
* https://hub.docker.com/r/jenkins/inbound-agent[Inbound agent]
* https://hub.docker.com/r/jenkins/ssh-agent[Outbound (ssh) agent]

Tool specific agent containers will no longer include Java 8:

* https://hub.docker.com/r/jenkins/jnlp-agent-docker[Docker agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-golang[Golang agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-maven[Maven agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-node[NodeJS agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-powershell[Powershell agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-python3[Python3 agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-python[Python agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-ruby[Ruby agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-terraform[Terraform agent]

> The Java 8 dedicated agent image will no longer be updated:

Reviewer 3: Do not care as much about leaving old agent images around, since the agent is generally untrusted anyway.

* Remove Java 8 from weekly core release, weekly Docker controller images
* Announce Java 8 end of support for LTS in blogpost
* Remove Java 8 from LTS core release, include in changelog and upgrade guide
* Remove Java 8 from Docker agent images

==== Website

* link:https://jenkins.io/doc/administration/requirements/java/[Java Support Page] is updated to state the weekly version of Jenkins core and the LTS version of Jenkins core that last support Java 8
* A blogpost is provided that announces the change in weekly releases and outlines the steps administrators must take to make the change
** War file installations
** Docker installations
** MSI installations on Windows
** RPM and DEB installations on Linux
* A blogpost is provided that announces the change in an LTS release and outlines the steps administrators must take to make the change
* The LTS changelog and upgrade guide describes the steps administrators must take to make the change
* A webinar is presented that outlines the changes and outlines the steps administrators must take to make the change

==== Issue tracking

* Issues related to Java 8 end of support are tracked as Jenkins issues

> ** link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688] is the Jira epic that tracks issues in Jira

Author: I've started the review of issues in those lists and have assigned a few of them to the [epic](https://issues.jenkins.io/browse/JENKINS-67688).  As I'm able to complete more reviews or as others are able to complete reviews, more items will be added.

** Plugins that use GitHub issues will place a link to their GitHub issue into the Jira epic

==== Post-release support

After the end of Java 8 support in the weekly releases, there may be a number of issues reported by early adopters.
Core maintainers will respond to issue reports as they did for configuration form modernization ("table to div").

> A Jira label `java8-end-of-support` will be assigned to issue reports related to Java 8 end of support.

Reviewer 3: We already have a jumble of labels all essentially about the same thing. I think this wording is wrong. There is no issue about ending Java 8 support per se, or at least none we intend to do anything about; issues would be that something is broken in Java 11, for which we already have labels, and that this has now become critical because it is impossible to update Jenkins core until the issue is addressed or worked around.

==== LTS Backporting

All backporting will be done according to the link:https://jenkins.io/download/lts/#backporting-process[LTS Backporting Process].

There is no plan to backport changes for the end of Java 8 support to previous LTS baselines.

== Motivation

Java 11 was released on September 25, 2018.
It is an LTS version with a long support timeline.
Java 8 was released in March 2014.
The link:https://www.oracle.com/java/technologies/java-se-support-roadmap.html[Oracle Java SE Support Roadmap] states that premier support for Java 8 ends in March 2022.
Java 8 enhancements have stopped as effort focuses on Java 11, Java 17, and beyond.

Removing support for Java 8 simplifies the supported configurations and allows further modernization of Jenkins core.

> Ending support for Java 8 allows Jenkins core and Jenkins plugins to use libraries that support Java 11 but do not support Java 8.

Reviewer 3: And directly use new language and API features in Java 11.

The main objective is to move Jenkins core development to Java 11.
There will be follow-up tasks for further improvements and to adopt new features.

=== Support of Java 17

This JEP intentionally limits its scope by not including Java 17 support.
It does not prevent work on Java 17, but that work is outside the scope of this JEP.

=== Docker image labeling

Docker image labels were updated in August 2021 to use link:https://www.jenkins.io/blog/2021/08/17/docker-images-use-jdk-11-by-default/[Java 11 by default].
The image labels that do not explicitly mention a Java version (like `latest`, `lts`, `slim`, `alpine`) are already delivering Java 11.

Image labels that explicitly mention `jdk8` will not be updated after Jenkins core ends support for Java 8.

== Backwards Compatibility

The following backward compatibility requirements are defined:

* Jenkins core and updated plugins should fully support Java 11
* Jenkins plugins may continue to compile with Java 8 so long as the plugins run successfully with Java 11
* Jenkins plugins that require a Jenkins version that does not support Java 8 will be expected to compile with Java 11

== Security

=== Process

> Only Java 11 with the latest security fixes will be supported at the moment of the public release.

Reviewer 3: ```suggestion
Only Java 11 with the latest security fixes will be supported at the moment of the first LTS release requiring Java 11.
```

Jenkins security issues on the release that ends support of Java 8 will be processed according to the

> standard link:https://jenkins.io/security/[Jenkins Security Process].

Reviewer 3: Meaning, discarded typically?

=== Security risks

* No additional security risks are expected due to Jenkins ending support for Java 8

== Infrastructure Requirements

=== ci.jenkins.io

* Tool Infrastructure should continue to offer the latest version of Java 11

=== Jenkins Pipeline Library

> * `buildPlugin()`, `runATH()`, and `runPCT()` will run tests with JDK 11

Reviewer 3: And `buildPlugin` will compile _using_ JDK 11, even if the plugin is still compatible with 8. `-release` should ensure that the plugin does not accidentally use post-8 APIs. It might however begin implicitly relying on behavioral changes between 8 and 11, resulting in a functional incompatibility with older core versions running on 8.

=== DockerHub

* Dockerhub will continue to host container images for Java 11

== Testing

Ending Java 8 support in Jenkins requires significant testing.

> Community contributors will be encouraged to test environments and configurations to assure that Jenkins core no longer requires Java 8.

Reviewer 3: Again I think this is better expressed in terms of requiring 11 and ensuring that everything runs smoothly on 11.

A link:https://docs.google.com/document/d/13ttjJ7HaUkYMy3L5P8D7w7TddqrUr-1IojtZCukFBQ8/edit?usp=sharing[status reporting document] is ready to track the testing effort.
Testers are welcome to report their results there.

Tests to be performed:

* ATH is updated and successful on Java 11 (**done** in link:https://github.com/jenkinsci/acceptance-test-harness/pull/726[ATH PR 726])
* PCT is updated and successful on Java 11

> * Plugin bill of materials is updated and successful on Java 11

Reviewer 3: Unclear what this means. Do you mean that the `bom` PCT tests pass on 11? If so, then this is the same as the previous bullet point, since AFAIK nothing else in Jenkins infra actually runs PCT.

* Packaging tests are performed successfully on Java 11

> * Exploratory tests are performed successfully to check for inadvertent use of Java 8

Reviewer 3: What does this mean?

Additional prototypes may be evaluated using pull requests or forks of Jenkins core.
These prototypes include Jenkins core, Docker updates and downstream demo patches.

* link:https://github.com/jenkinsci/jenkins/pull/6092[Announce forthcoming Java 8 EOL]
* link:https://github.com/jenkinsci/jenkins/pull/6083[Drop core support for Java 8]
* link:https://github.com/jenkins-infra/helpdesk/issues/2758#issuecomment-1018670240[Update infrastructure JDK versions]
* link:https://github.com/jenkinsci/docker/blob/master/.github/dependabot.yml[Dependabot updates for controller images]
* link:https://github.com/jenkinsci/docker-agent/blob/master/.github/dependabot.yml[Dependabot updates for agent images]
* link:https://github.com/jenkinsci/docker-inbound-agent/blob/master/.github/dependabot.yml[Dependabot updates for inbound agent images]
* link:https://github.com/jenkinsci/docker-ssh-agent/blob/master/.github/dependabot.yml[Dependabot updates for outbound (ssh) agent images]

== References

* link:https://www.oracle.com/java/technologies/java-se-support-roadmap.html[Oracle Java SE Support Roadmap]
* link:https://jenkins.io/doc/administration/requirements/java/[Java requirements] in Jenkins

> * link:[Java 8 end of support testing status document]

Reviewer 3: Did you forget the link target?

= JEP-0000: End Java 8 support in Jenkins
:toc: preamble
:toclevels: 3
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

.Metadata
[cols="1h,1"]
|===
| JEP
| 0000

| Title
| End Java 8 support in Jenkins

| Sponsor
| link:https://github.com/MarkEWaite[Mark Waite]

// Use the script `set-jep-status <jep-number> <status>` to update the status.
| Status
| Not Submitted :information_source:

| Type
| Process

| Created
| 2022-01-28

| BDFL-Delegate
| TBD

| Pull requests
|

> - link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core

Reviewer: maybe? it's not rendering well 
```suggestion

- link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core
```

= JEP-0000: End Java 8 support in Jenkins
:toc: preamble
:toclevels: 3
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

.Metadata
[cols="1h,1"]
|===
| JEP
| 0000

| Title
| End Java 8 support in Jenkins

| Sponsor
| link:https://github.com/MarkEWaite[Mark Waite]

// Use the script `set-jep-status <jep-number> <status>` to update the status.
| Status
| Not Submitted :information_source:

| Type
| Process

| Created
| 2022-01-28

| BDFL-Delegate
| TBD

| Pull requests
|

- link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core

> - link:https://github.com/jenkinsci/jenkins/pull/6086[PR 6086] - Switch Jenkins acceptance test harness in core to Java 11

Reviewer: hmm still doesn't work

Author: Rendering fixed in deb9bf19696a4386e0558aab05f449c45ac429c4

> - link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core
> - link:https://github.com/jenkinsci/jenkins/pull/6086[PR 6086] - Switch Jenkins acceptance test harness in core to Java 11
> - link:https://github.com/jenkinsci/jenkins/pull/6092[PR 6092] - Add administrative monitor to remind of upcoming end of Java 8 support
> - link:https://github.com/jenkinsci/acceptance-test-harness/pull/726[PR 726] - Drop Java 8 from acceptance test harness

Reviewer 4: ```suggestion
* link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core
* link:https://github.com/jenkinsci/jenkins/pull/6086[PR 6086] - Switch Jenkins acceptance test harness in core to Java 11
* link:https://github.com/jenkinsci/jenkins/pull/6092[PR 6092] - Add administrative monitor to remind of upcoming end of Java 8 support
* link:https://github.com/jenkinsci/acceptance-test-harness/pull/726[PR 726] - Drop Java 8 from acceptance test harness
```

Author: Fixed in deb9bf1 by enabling asciidoc parsing inside the table cell.

//| Resolution
//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:

|===

== Abstract

Java 11 was released on September 25, 2018.
It is an LTS version with a long support timeline.
Java 8 was released in March 2014.
The link:https://www.oracle.com/java/technologies/java-se-support-roadmap.html[Oracle Java SE Support Roadmap] states that premier support for Java 8 ends in March 2022.
Java 8 enhancements have stopped as effort focuses on Java 11, Java 17, and beyond.

This Jenkins Enhancement Proposal describes actions required
to make Jenkins end its support of Java 8.

== Specification

The specification has been created as a result of a link:https://groups.google.com/g/jenkinsci-dev/c/YghQ0YP4m78/m/LO9AFa_GAgAJ[Jenkins developers mailing list discussion].

All stories in this JEP have tickets created in Jira or pull requests in GitHub.
Jira tickets are aggregated in
link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688].

If the tickets are not linked in the description, they can be found there.

=== Goals and non-goals

Goals:

* Jenkins core development requires Java 11 or newer

Non-goals:

> * Jenkins plugin development requires Java 11 or newer

Author: I think you're right.

What if a new goal were added to state that plugin development requires Java 11 when the Jenkins minimum version supported by the plugin requires Java 11?

I'm trying to state that until a plugin requires a Jenkins version that does not support Java 8, that plugin is not strictly required to be built with Java 11.  Once the plugin requires a Jenkins version that does not support Java 8, the plugin must compile with Java 11 or newer.

> * Jenkins plugin development requires Java 11 or newer

Reviewer: sure

> * Jenkins plugin development requires Java 11 or newer

Author: Fixed in f54d9d4

  They will be handled in follow-up JEPs if needed.
* Gradle build flow

=== Jenkins core patches

Work to be considered is defined in link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688].

==== Library updates

* The link:https://issues.jenkins.io/browse/JENKINS-67688[JENKINS-67688 epic] will include library updates as needed.
* Some updates may require downstream plugin updates.

==== Core patches

* link:https://github.com/jenkinsci/jenkins/pull/6083[PR 6083] - Drop Java 8 from Jenkins core
* link:https://github.com/jenkinsci/jenkins/pull/6086[PR 6086] - Switch Jenkins acceptance test harness in core to Java 11
* link:https://github.com/jenkinsci/jenkins/pull/6092[PR 6092] - Add administrative monitor to remind of upcoming end of Java 8 support

==== Build flow updates

* Jenkinsfile is updated to stop running tests with Java 8
** It includes Unit tests, JTH and ATH smoke tests
* It is possible to build Jenkins Core with the release profile on Java 8

=== Jenkins Docker packaging

The containers tagged for Java 8, like `latest-jdk8` and `centos7-jdk8` will no longer be updated.

> The upgrade guide and the announcement blogpost will note that users must switch to other images.

Reviewer 4: As a final label, the docker image could include a warning to tell users to switch to another label.

Author: Good suggestion.  I'm not sure how that would best be implemented.  

Would the Docker image add a new "flag file" to Jenkins home directory (this image is no longer being updated) and then the presence of that flag file would cause an administrative monitor to be displayed?

Is there a better way to get the message to the Jenkins administrator?

Reviewer 4: I just thought about a plain warning message in stderr, but indeed, that could go further and be implemented as an administrative monitor as you suggest.

Reviewer: there's already an admin monitor to tell them to switch and a pending PR to add an EOL date

Author: > I just thought about a plain warning message in stderr, but indeed, that could go further and be implemented as an administrative monitor as you suggest.

I fear that warnings to the stderr of a Jenkins controller will not be seen by most administrators.

Author: > there's already an admin monitor to tell them to switch and a pending PR to add an EOL date

I was thinking that this admin monitor would be informing the user that the Docker image is no longer being updated.  We could have used that for the end of life of the centos8 image and would need something similar for the end of life for the Debian 10 (buster) image when Debian 12 is released.

I guess that an admin monitor may not be well suited for that idea, since we would want a distinct admin monitor for each image end of life.  If we tried to use the same admin monitor for all image end of life messages, then the administrator may have dismissed the end of life for an earlier image and we need to show them that the current image they are using has reached the end of its life

Reviewer 3: OK. Just not sure how to make it clear to people consuming images that they are out of support.

Author: > OK. Just not sure how to make it clear to people consuming images that they are out of support

I think you're right that we have a problem there.  We don't have a way to tell users that the Docker image they are using is no longer being updated.  Maybe we should consider a new administrative monitor each time we stop updates on a Docker container.  If Jenkins is running in a container that is no longer being updated, we tell the administrator at least once that they need to move to a different Docker label.

Java 8 images will no longer be provided for the controller containers:

* https://hub.docker.com/r/jenkins/jenkins[Controller]

==== Jenkins Agent Images

Java 8 images will no longer be provided for the general purpose agent containers:

* https://hub.docker.com/r/jenkins/agent[Agent]
* https://hub.docker.com/r/jenkins/inbound-agent[Inbound agent]
* https://hub.docker.com/r/jenkins/ssh-agent[Outbound (ssh) agent]

> Tool specific agent containers will no longer include Java 8:

Author: Agreed.  They seem to be updated regularly with the latest base image, but only have tags for `latest`.

I didn't put text into this about the state of those images because I thought it was a separate topic.

* https://hub.docker.com/r/jenkins/jnlp-agent-docker[Docker agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-golang[Golang agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-maven[Maven agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-node[NodeJS agent]
* https://hub.docker.com/r/jenkins/jnlp-agent-powershell[Powershell agent]

> * https://hub.docker.com/r/jenkins/jnlp-agent-python2[Python2 agent]

Author: Thanks for catching that.  Fixed in 5615e48

The Docker containers are already using Java 11.

=== Plugins

No updates are expected to be required in plugins for this JEP.
Plugins compiled with Java 8 are expected to continue running with Jenkins core compiled with Java 11.
Incompatibilities will be reported and tracked as plugin issue reports.

=== New policy: Jenkins core with Java 11

The following policy is suggested:

* Jenkins core components will be compiled with Java 11 and will require Java 11 or later at runtime
* Jenkins plugins that depend on a Jenkins core that requires Java 11 must be compiled with Java 11
** In order to support releases that only run with Java 11, the plugins must use the plugin POM that adds support for `java.level` 11

This policy may require patches in parent POMs:

* Parent POMs will be updated as needed, including
  link:https://github.com/jenkinsci/plugin-pom[Plugin POM],
  link:https://github.com/jenkinsci/pom[Jenkins POM], and
  link:https://github.com/jenkinsci/bom[Jenkins plugin bill of materials POM],

=== Rollout plan

The rollout procedure should be coordinated within the link:https://jenkins.io/sigs/platform/[Platform SIG].

> ==== Timeline

Author: Correct.  That entire timeline section should have been replaced with "TBD" while working through the rest of the plan.  Will fix that.

to ensure it can be delivered by a small team.
Non-goals in the specification are defined to limit the scope of work.
The main objective is to move Jenkins core development to Java 11.
There will be follow-up tasks for further improvements and to adopt new features.

=== Support of Java 17

This JEP intentionally limits its scope by not including Java 17 support.
It does not prevent work on Java 17, but that work is outside the scope of this JEP.

=== Docker image labeling

Docker image labels were updated in August 2021 to use link:https://www.jenkins.io/blog/2021/08/17/docker-images-use-jdk-11-by-default/[Java 11 by default].
The image labels that do not explicitly mention a Java version (like `latest`, `lts`, `slim`, `alpine`) are already delivering Java 11.

Image labels that explicitly mention `jdk8` will not be updated after Jenkins core ends support for Java 8.

== Backwards Compatibility

The following backward compatibility requirements are defined:

* Jenkins core and updated plugins should fully support Java 11
* Jenkins plugins may continue to compile with Java 8 so long as the plugins run successfully with Java 11

> * Jenknis plugins that require a Jenkins version that does not support Java 8 will be expected to compile with Java 11

Reviewer: ```suggestion
* Jenkins plugins that require a Jenkins version that does not support Java 8 will be expected to compile with Java 11
```

== Security

=== Process

Only Java 11 with the latest security fixes will be supported at the moment of the public release.

Jenkins security issues on the release that ends support of Java 8 will be processed according to the
standard link:https://jenkins.io/security/[Jenkins Security Process].

=== Security risks

* No additional security risks are expected due to Jenkins ending support for Java 8

== Infrastructure Requirements

=== ci.jenkins.io

* Tool Infrastructure should continue to offer the latest version of Java 11

=== Jenkins Pipeline Library

* `buildPlugin()`, `runATH()`, and `runPCT()` will run tests with JDK 11

=== DockerHub

* Dockerhub will continue to host container images for Java 11

== Testing

Ending Java 8 support in Jenkins requires significant testing.
Community contributors will be encouraged to test environments and configurations to assure that Jenkins core no longer requires Java 8.

A link:https://docs.google.com/document/d/13ttjJ7HaUkYMy3L5P8D7w7TddqrUr-1IojtZCukFBQ8/edit?usp=sharing[status reporting document] is ready to track the testing effort.
Testers are welcome to report their results there.

Tests to be performed:

> * ATH is updated as needed to run on Java 11

Reviewer: this was completed in https://github.com/jenkinsci/acceptance-test-harness/pull/726

Author: Fixed in d3320872d7d5091aa43e1f47f289970338fe1b81

> * ATH is performed successfully on Java 11

Reviewer: completed it in https://github.com/jenkinsci/acceptance-test-harness/pull/726

Author: Fixed in d3320872d7d5091aa43e1f47f289970338fe1b81

* PCT is updated as needed to run on Java 11
* PCT is performed successfully on Java 11

> * Packaging tests are performed successfully on Java 11

Reviewer: I think @dduportal might have already done this

Reviewer 5: Not yet (ref. https://github.com/jenkins-infra/release/issues/199) but JDK11 is available on the current release process so easy to switch.
We can plan to do it for the 1st time for the weekly of the 7th of feb.(or later).

* Exploratory tests are performed successfully to check for inadvertent use of Java 8

== Prototype Implementation

> Prototype implementations have been created by pull requests.

Reviewer: this line doesn't really make sense to me

Author: Removed in d3320872d7d5091aa43e1f47f289970338fe1b81
